### ISSUE TITLE
IGC Valgrind issues

### ISSUE THREAD

**Issue by jeffhammond** [intel/intel-graphics-compiler#150]
I am trying to debug a simple GPU application with Valgrind but IGC generates many diagnostic messages, which makes it quite difficult to debug my application code.  Valgrind is an essential tool for HPC debugging and IGC should not make it harder to use.

Numerous instances of the following diagnostic are produced when IGC is used.  Can you determine if these are false positives or not?

```
==2548== Mismatched free() / delete / delete []
==2548==    at 0x4C3123B: operator delete(void*) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==2548==    by 0x202B3376: llvm::GVN::ValueTable::clear() (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x202B373F: llvm::GVN::cleanupGlobalSets() (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x202C24F8: llvm::GVN::runImpl(llvm::Function&, llvm::AssumptionCache&, llvm::DominatorTree&, llvm::TargetLibraryInfo const&, llvm::AAResults&, llvm::MemoryDependenceResults*, llvm::LoopInfo*, llvm::OptimizationRemarkEmitter*) (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x202C2DF1: ??? (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x20ABA7C1: llvm::FPPassManager::runOnFunction(llvm::Function&) (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x20ABB1C0: llvm::FPPassManager::runOnModule(llvm::Module&) (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x20ABB63B: llvm::legacy::PassManagerImpl::run(llvm::Module&) (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x1F6A5BFC: ??? (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x1F53160C: ??? (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x1F5EE4C5: ??? (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x15F9106B: NEO::CompilerInterface::build(NEO::Device const&, NEO::TranslationInput const&, NEO::TranslationOutput&) (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libze_intel_gpu.so.1.0.17767)
==2548==  Address 0x27e90e80 is 0 bytes inside a block of size 8,192 alloc'd
==2548==    at 0x4C320A6: memalign (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==2548==    by 0x15A4C45A: operator new(unsigned long, std::align_val_t) (new_opa.cc:145)
==2548==    by 0x1F620819: ??? (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x202B7081: ??? (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x202B8528: llvm::GVN::ValueTable::lookupOrAdd(llvm::Value*) (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x202C0CFE: llvm::GVN::processInstruction(llvm::Instruction*) (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x202C14C7: llvm::GVN::processBlock(llvm::BasicBlock*) (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x202C228B: llvm::GVN::iterateOnFunction(llvm::Function&) (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x202C24C4: llvm::GVN::runImpl(llvm::Function&, llvm::AssumptionCache&, llvm::DominatorTree&, llvm::TargetLibraryInfo const&, llvm::AAResults&, llvm::MemoryDependenceResults*, llvm::LoopInfo*, llvm::OptimizationRemarkEmitter*) (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x202C2DF1: ??? (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x20ABA7C1: llvm::FPPassManager::runOnFunction(llvm::Function&) (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
==2548==    by 0x20ABB1C0: llvm::FPPassManager::runOnModule(llvm::Module&) (in /opt/intel/oneapi/OpenCL/beta09_18.04LTS/libigc.so.1.0.4756)
```

**Comment by eero-t** [comment]
Because this is about standard alloc/free function mismatches, I don't see how Valgrind could get that wrong.

(GPU allocs would be a different matter, then there would be also a question of whether all the related components have added annotations for those, and are they built with Valgrind headers present so that those annotations would actually be built-in.)

**Comment by paigeale** [comment]
Issue has been tested with a newer igc version release https://github.com/intel/intel-graphics-compiler/releases/tag/igc-1.0.7423. Please reopen if issue persists and we will have someone relook into this
